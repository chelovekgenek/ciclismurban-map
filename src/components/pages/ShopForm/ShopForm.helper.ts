import { MapOptions } from "helpers"
import { ShopModel } from "models/location"

export class ShopFormValues extends ShopModel {
  constructor(props?: ShopModel) {
    super()

    if (props) {
      this.title = props.title
      this.description = props.description
      this.point = props.point
      this.image = props.image
    }
  }

  title = ""
  description = ""
  point = MapOptions.position
  image = ""
}
