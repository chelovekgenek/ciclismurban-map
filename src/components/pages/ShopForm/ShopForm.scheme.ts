import { ShopModel, WeeklyScheduleModel } from "@ciclismurban/models"
import { MapOptions } from "helpers"

export class ShopFormValues extends ShopModel {
  constructor(props?: ShopModel) {
    super()

    if (props) {
      this.title = props.title
      this.description = props.description
      this.point = props.point
      this.image = props.image
      this.schedule = props.schedule
    }
  }

  title = ""
  description = ""
  point = MapOptions.position
  image = ""
  schedule: WeeklyScheduleModel = {
    mon: [],
    tue: [],
    wed: [],
    thu: [],
    fri: [],
    sat: [],
    sun: [],
  }
}
