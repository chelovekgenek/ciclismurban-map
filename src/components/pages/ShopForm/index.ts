export { ShopFormContainer as ShopForm } from "./ShopForm.container"
