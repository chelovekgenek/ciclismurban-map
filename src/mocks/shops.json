[{ "point": { "lat": 47.0265823, "lng": 28.8180033 } }]
