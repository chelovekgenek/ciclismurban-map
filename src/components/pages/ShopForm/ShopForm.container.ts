import { connect } from "react-redux"
import { withFormik } from "formik"
import { withRouter } from "react-router-dom"
import { ShopModel, LocationExposeGroup } from "@ciclismurban/models"

import { SelectedActions, getSelectedRoot, SelectedGetActions } from "store/entities/locations"
import { Actions as ShopsActions, Selectors as ShopsSelectors } from "store/entities/locations/shops"
import { TAppState } from "store/entities"
import { handleLocationFormSubmit, getHandlerLocationFormValidate } from "helpers"

import { IProps, ShopForm } from "./ShopForm"
import { ShopFormValues } from "./ShopForm.scheme"

export type IStateProps = {
  shops: {
    fetching: ReturnType<typeof ShopsSelectors.getFetching>
  }
  selected: ReturnType<typeof getSelectedRoot>
}

export interface IDispatchProps {
  create: typeof ShopsActions.Create.request
  update: typeof ShopsActions.Update.request
  getSelected: typeof SelectedGetActions.request
  clearSelected: typeof SelectedActions.clear
}

const hocConnect = connect<IStateProps, IDispatchProps, {}, TAppState>(
  state => ({
    shops: {
      fetching: ShopsSelectors.getFetching(state),
    },
    selected: getSelectedRoot(state),
  }),
  {
    create: ShopsActions.Create.request,
    update: ShopsActions.Update.request,
    getSelected: SelectedGetActions.request,
    clearSelected: SelectedActions.clear,
  },
)
const hocWithFormik = withFormik<IProps, ShopFormValues>({
  enableReinitialize: true,
  mapPropsToValues: props => new ShopFormValues(props.selected.data as ShopModel),
  handleSubmit: handleLocationFormSubmit,
  validate: getHandlerLocationFormValidate(ShopModel, [LocationExposeGroup.WRITE]),
  validateOnBlur: true,
})

export const ShopFormContainer = withRouter(hocConnect(hocWithFormik(ShopForm)))
