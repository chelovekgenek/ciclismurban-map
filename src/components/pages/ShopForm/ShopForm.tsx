import React, { useEffect, useMemo } from "react"
import { Form, FormikProps } from "formik"
import { RouteComponentProps } from "react-router-dom"
import { isEmpty } from "lodash-es"

import { App } from "components/generic/layout"
import { Input, Textarea, GoogleMap, Upload } from "components/generic/form"
import { Button } from "components/generic/ui"

import { IStateProps, IDispatchProps } from "./ShopForm.container"
import { ShopFormValues } from "./ShopForm.helper"

interface RouteParams {
  id: string
}
export interface IProps extends IStateProps, IDispatchProps, RouteComponentProps<RouteParams> {}

export const ShopForm: React.FC<IProps & FormikProps<ShopFormValues>> = ({
  submitForm,
  resetForm,
  isValid,
  touched,
  match,
  selected,
  getSelected,
  clearSelected,
  shops,
}) => {
  useEffect(() => {
    if (match.params.id) {
      getSelected({ entity: "shops", uuid: match.params.id })
    }
    return () => {
      clearSelected()
    }
  }, [match.params.id, getSelected, clearSelected])

  const actions = useMemo(() => {
    const reset = (
      <Button key="1" onClick={() => resetForm(selected.data as ShopFormValues)} disabled={isEmpty(touched)}>
        Сбросить
      </Button>
    )
    const submit = (
      <Button key="2" type="primary" onClick={submitForm} disabled={!isValid} loading={shops.fetching}>
        Сохранить
      </Button>
    )
    return selected.data ? [reset, submit] : [submit]
  }, [selected.data, shops.fetching, isValid, submitForm, resetForm, touched])
  const title = useMemo(() => (selected.data ? "Редактировать магазин" : "Новый магазин"), [selected.data])

  return (
    <App
      content={{
        useHeader: true,
        useLayout: true,
        title,
        actions,
      }}
    >
      <Form>
        <Input label="Название" name="title" />
        <Textarea label="Описание" name="description" autosize={{ minRows: 1, maxRows: 4 }} />
        <Upload label="Изображение" name="image" />
        <GoogleMap name="point" />
      </Form>
    </App>
  )
}
