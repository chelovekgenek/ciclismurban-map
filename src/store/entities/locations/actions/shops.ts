import { IState as IShopsState } from "../reducers/shops"
import { action, payload } from "ts-action"

export enum ShopsTypes {
  GET__REQUEST = "LOCATIONS__SHOPS__GET__REQUEST",
  GET__SUCCESS = "LOCATIONS__SHOPS__GET__SUCCESS",
  GET__FAILURE = "LOCATIONS__SHOPS__GET__FAILURE",
}

type TRequiredShopsState = Required<IShopsState>
export const ShopsActions = {
  requestGet: action(ShopsTypes.GET__REQUEST),
  successGet: action(ShopsTypes.GET__SUCCESS, payload<TRequiredShopsState["data"]>()),
  failureGet: action(ShopsTypes.GET__FAILURE, payload<TRequiredShopsState["error"]>()),
}
