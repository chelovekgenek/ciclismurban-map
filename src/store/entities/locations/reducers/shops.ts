import { reducer, on, union } from "ts-action"
import { ShopModel } from "@ciclismurban/models"

import { ShopsGetActions, ShopsCreateActions, ShopsUpdateActions, ShopsDeleteActions } from "../actions"

export interface IState {
  data: ShopModel[]
  fetching: boolean
  error?: Error
}

const initialState: IState = {
  data: [],
  fetching: false,
}

export default reducer(
  initialState,
  on(
    ...union(
      ShopsGetActions.request,
      ShopsCreateActions.request,
      ShopsUpdateActions.request,
      ShopsDeleteActions.request,
    ),
    state => ({
      ...state,
      fetching: true,
    }),
  ),
  on(
    ...union(
      ShopsGetActions.failure,
      ShopsCreateActions.failure,
      ShopsUpdateActions.failure,
      ShopsDeleteActions.failure,
    ),
    (state, { payload }) => ({
      ...state,
      fetching: false,
      error: payload,
    }),
  ),
  on(ShopsGetActions.success, (state, { payload }) => ({ ...state, fetching: false, data: payload })),
  on(ShopsCreateActions.success, (state, { payload }) => ({
    ...state,
    fetching: false,
    data: state.data.concat(payload),
  })),
  on(ShopsUpdateActions.success, (state, { payload }) => ({
    ...state,
    fetching: false,
    data: state.data.map(item => (item.uuid === payload.uuid ? payload : item)),
  })),
  on(ShopsDeleteActions.success, (state, { payload }) => ({
    ...state,
    data: state.data.filter(item => item.uuid !== payload),
    fetching: false,
  })),
)
