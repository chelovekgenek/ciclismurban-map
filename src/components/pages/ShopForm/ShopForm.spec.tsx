import React from "react"
import { shallow } from "enzyme"

import formProps from "mocks/formik-form-props"
import shops from "mocks/shops.json"

import { ShopFormValues } from "./ShopForm.scheme"
import { ShopForm } from "./ShopForm"

describe("pages/ShopForm", () => {
  const initialValues = new ShopFormValues()
  const render = (props: Partial<React.ComponentProps<typeof ShopForm>> = {}) =>
    shallow(
      <ShopForm
        {...formProps}
        initialValues={initialValues}
        values={initialValues}
        shops={{ fetching: false }}
        selected={{ fetching: false }}
        getSelected={jest.fn() as any}
        clearSelected={jest.fn() as any}
        create={jest.fn() as any}
        update={jest.fn() as any}
        match={{ params: {} } as any}
        history={{} as any}
        location={{} as any}
        {...props}
      />,
    )
  it("should match snapshot", () => {
    expect(render()).toMatchSnapshot()
  })
  it("should match snapshot if `selected.data` is provided", () => {
    expect(render({ selected: { data: shops[0] } as any })).toMatchSnapshot()
  })
  it("should retreive dataset by `match.params.id`", () => {
    // TODO https://github.com/airbnb/enzyme/issues/2091
    // const props = {
    //   match: { params: { id: "id" } } as any,
    //   getSelected: jest.fn() as any,
    // }
    // render(props)
    // expect(props.getSelected).toHaveBeenCalledWith({ entity: "shops" })
  })
  it("should be able to reset form if touched", () => {
    const wrapper = render({ selected: { data: shops[0] } as any, touched: { title: true } })
    expect((wrapper.find("App").prop("content") as any).actions[0].props.disabled).toBe(false)
  })
})
