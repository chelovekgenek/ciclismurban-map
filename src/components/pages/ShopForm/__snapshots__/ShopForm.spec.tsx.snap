// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/ShopForm should match snapshot 1`] = `
<App
  content={
    Object {
      "actions": Array [
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[MockFunction]}
          type="primary"
        >
          Сохранить
        </Button>,
      ],
      "title": "Новый магазин",
      "useHeader": true,
      "useLayout": true,
    }
  }
>
  <Form>
    <Input
      label="Название"
      name="title"
    />
    <Textarea
      autosize={
        Object {
          "maxRows": 4,
          "minRows": 1,
        }
      }
      label="Описание"
      name="description"
    />
    <Upload
      label="Изображение"
      name="image"
    />
    <GoogleMap
      name="point"
    />
  </Form>
</App>
`;

exports[`pages/ShopForm should match snapshot if \`selected.data\` is provided 1`] = `
<App
  content={
    Object {
      "actions": Array [
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
        >
          Сбросить
        </Button>,
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[MockFunction]}
          type="primary"
        >
          Сохранить
        </Button>,
      ],
      "title": "Редактировать магазин",
      "useHeader": true,
      "useLayout": true,
    }
  }
>
  <Form>
    <Input
      label="Название"
      name="title"
    />
    <Textarea
      autosize={
        Object {
          "maxRows": 4,
          "minRows": 1,
        }
      }
      label="Описание"
      name="description"
    />
    <Upload
      label="Изображение"
      name="image"
    />
    <GoogleMap
      name="point"
    />
  </Form>
</App>
`;
